modules/local/locidex_select.nf

```diff
@@ -0,0 +1,276 @@
+/*
+Locidex provides the option to select a database from a group of databases, by
+using a manifest file. To prevent copying all databases each time an allele database
+needs to be selected this file will read only the "manifest" file of the databases and
+to pick the correct allele scheme.
+
+
+Locidex Manifest is setup as:
+    {
+        db_name: [
+            {
+                "path": path/to/db/relative/to/manifest
+                # DB Config data, the newest db data will be selected as versions are not standardized
+                "config": {
+                    "db_name": "Locidex Database 1",
+                    "db_version": "1.0.0",
+                    "db_date": "yyyy-MM-dd",
+                    "db_author": "test1",
+                    "db_desc": "test1",
+                    "db_num_seqs": 53,
+                    "is_nucl": true,
+                    "is_prot": true,
+                    "nucleotide_db_name": "nucleotide",
+                    "protein_db_name": "protein"
+                }
+            }
+        ]
+    }
+*/
+
+import groovy.json.JsonSlurper
+import groovy.json.JsonBuilder
+import java.text.SimpleDateFormat
+
+
+process LOCIDEX_SELECT {
+    tag "$meta.id"
+    label "process_single"
+
+    input:
+    tuple val(meta), val(top_hit), val(contigs)
+    val manifest // This is a json file to be parsed
+    val config // optional path to a database file
+
+    output:
+    tuple val(meta), val(contigs), val(scheme), val(paired_p), emit: db_data
+    tuple val(meta), path(output_config), emit: config_data
+
+    exec:
+
+    report_name = "${meta.id}_${params.locidex.db_config_output_name}"
+    output_config = task.workDir.resolve(report_name)
```

Comment:
This is just to create the variable initially as it is ued in both of the control flow branches. For clarification however the `output_config` variable is decalred without a `def` declaration as it is not visible in the outer scope of the exec block. So when the variable is expaned it does so with the correct value.

Reply:
Thanks so much. I think this question and other questions related to using `val` instead of `path` for `exec` blocks comes from my own misunderstanding of Nextflow.

According to here https://groups.google.com/g/nextflow/c/tp_b1p0DBE4/m/2tRi2FAj26gJ, it seems that in an `exec` block, Nextflow is not changing to a separate work directory and so using `task.workDir.resolve` is recommended. Though that post is from 2015, so things might have changed (hard to find information about `exec` blocks though).

The above thread and this one https://github.com/nextflow-io/nextflow/issues/1678#issuecomment-927913406 also contains more information about why using `val` works for input files in an `exec` block, but why `path` doesn't work.

It seems that for `exec` blocks, the expectation was to do things like internal channel manipulation, and it was less designed for file manipulations. Since in your case you want to use `exec` to also access Nextflow variables and run Groovy code, it makes sense to use `exec` and use these workarounds for writing back output files like `task.workDir.resolve` and using `val` instead of `path` for input/output files.

This is all to say I don't think there is any needed changes here or elsewhere were I made comments about using `path` instead of `val`. Sorry about my own misunderstanding of Nextflow.

Reply:
No worries! I feel like the utility of `exec` functions should be clarified better in the docs. As it would go great lengths in helping Nextflow users adopt more Groovy instead of relying solely on Python.